onnxruntime/core/providers/cpu/math/einsum_utils.cc

```diff
@@ -0,0 +1,724 @@
+#include "einsum_utils.h"
+
+namespace onnxruntime {
+
+namespace EinsumOp {
+
+// This helps decide if we need to apply (and pay the cost) of a Transpose
+static bool IsTransposeRequired(size_t input_rank, const std::vector<size_t>& permutation) {
+  ORT_ENFORCE(input_rank == permutation.size(), "The rank of the input must match permutation size for Transpose");
+  
+  // No transpose required for scalars
+  if (input_rank == 0) {
+    return false; 
+  }
+
+  // Weeds out cases where permutation is something like [0, 1, 2] for a 3D input and so on
+  bool transpose_required = false;
+  for (size_t i = 0; i < input_rank; ++i) {
+    if (permutation[i] != i) {
+      transpose_required = true;
+      break;
+    }
+  }
+
+  return transpose_required;
+}
+     // We have an the result in an output "candidate". Now we have to copy the contents in its buffer
+// into the buffer of the actual output given to us by the execution frame
+// We need to do this because the buffer owned by the output tensor of the op could be user provided buffer
+static void CopyOutputCandidateIntoOpOutout(Tensor& output, Tensor& candidate) {
+  ORT_ENFORCE(output.SizeInBytes() == candidate.SizeInBytes(),
+              "Einsum op: The candidate output does not match the actual output's shape");
+  // There are no string tensors - so safely use memcpy
+  memcpy(output.MutableDataRaw(), candidate.DataRaw(), candidate.SizeInBytes());
+}
+// Here we take a "candidate output"(candidate output is a tensor that is a permutation and / or a reshape away from the final output),
+// and after a few operations to get it to the required output structure, copy it to the op's output
+template <typename T>
+static void FinalizeOutput(Tensor& candidate_output, const std::vector<int64_t>& subscript_indices_in_candidate,
+                           const std::vector<int64_t>& subscript_indices_to_output_indices,
+                           Tensor& output, const std::vector<int64_t>& output_dims, const AllocatorPtr& allocator) {
+  auto output_rank = output_dims.size();
+  ORT_ENFORCE(output_rank == subscript_indices_in_candidate.size());
+
+  // Identtify the permutation required by the op's output
+  std::vector<size_t> output_permutation;
```

Comment:
nit: identify. 

FWIW this is super helpful: [https://marketplace.visualstudio.com/items?itemName=EWoodruff.VisualStudioSpellCheckerVS2017andLater](https://marketplace.visualstudio.com/items?itemName=EWoodruff.VisualStudioSpellCheckerVS2017andLater)

Reply:
Fixed.